File: python/102/patterns/command1.py
# Command Interface
class Command:
    def execute(self):
        pass

    def undo(self):
        pass

# Receiver
class TV:
    def turn_on(self):
        print("TV is ON")

    def turn_off(self):
        print("TV is OFF")

# Concrete Commands
class TurnOnCommand(Command):
    def __init__(self, tv):
        self.tv = tv

    def execute(self):
        self.tv.turn_on()

    def undo(self):
        self.tv.turn_off()

class TurnOffCommand(Command):
    def __init__(self, tv):
        self.tv = tv

    def execute(self):
        self.tv.turn_off()

    def undo(self):
        self.tv.turn_on()

# Invoker
class RemoteControl:
    def __init__(self):
        self.command = None

    def set_command(self, command):
        self.command = command

    def press_button(self):
        if self.command:
            self.command.execute()

    def press_undo(self):
        if self.command:
            self.command.undo()

# Client Code
tv = TV()

turn_on = TurnOnCommand(tv)
turn_off = TurnOffCommand(tv)

remote = RemoteControl()

# Turn the TV ON
remote.set_command(turn_on)
remote.press_button()

# Undo action
remote.press_undo()

# Turn the TV OFF
remote.set_command(turn_off)
remote.press_button()

# Undo action
remote.press_undo()

File: typescript-basic/classes/command1.ts
// Command Interface
interface Command {
    execute(): void;
    undo(): void;
}

// Receiver
class ShoppingCart {
    private items: string[] = [];

    addItem(item: string): void {
        this.items.push(item);
        console.log(`Added ${item} to the cart. Current items: [${this.items.join(', ')}]`);
    }

    removeItem(item: string): void {
        const index = this.items.indexOf(item);
        if (index !== -1) {
            this.items.splice(index, 1);
            console.log(`Removed ${item} from the cart. Current items: [${this.items.join(', ')}]`);
        } else {
            console.log(`Item ${item} not found in the cart.`);
        }
    }
}

// Concrete Commands
class AddItemCommand implements Command {
    constructor(private cart: ShoppingCart, private item: string) {}

    execute(): void {
        this.cart.addItem(this.item);
    }

    undo(): void {
        this.cart.removeItem(this.item);
    }
}

class RemoveItemCommand implements Command {
    constructor(private cart: ShoppingCart, private item: string) {}

    execute(): void {
        this.cart.removeItem(this.item);
    }

    undo(): void {
        this.cart.addItem(this.item);
    }
}

// Invoker
class CommandManager {
    private history: Command[] = [];

    executeCommand(command: Command): void {
        command.execute();
        this.history.push(command);
    }

    undoLastCommand(): void {
        if (this.history.length > 0) {
            const lastCommand = this.history.pop();
            lastCommand?.undo();
        } else {
            console.log("Nothing to undo!");
        }
    }
}

// Client Code
const cart = new ShoppingCart();
const commandManager = new CommandManager();

// Create commands
const addApple = new AddItemCommand(cart, "Apple");
const addBanana = new AddItemCommand(cart, "Banana");
const removeApple = new RemoveItemCommand(cart, "Apple");

// Execute commands
commandManager.executeCommand(addApple);    // Add Apple
commandManager.executeCommand(addBanana);  // Add Banana
commandManager.executeCommand(removeApple); // Remove Apple

// Undo commands
commandManager.undoLastCommand(); // Undo remove Apple
commandManager.undoLastCommand(); // Undo add Banana
commandManager.undoLastCommand(); // Undo add Apple
commandManager.undoLastCommand(); // Nothing to undo
